source/sigv4_quicksort.c

```diff
@@ -0,0 +1,171 @@
+ */
+static void quickSortHelper( void * pArray,
+                             size_t low,
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator );
+
+/**
+ * @brief A helper function to partition a subarray using the last element
+ * of the array as the pivot. All items smaller than the pivot end up
+ * at its left while all items greater than end up at its right.
+ *
+ * @param[in] pArray The array to be sorted.
+ * @param[in] low The low index of the array.
+ * @param[in] high The high index of the array.
+ * @param[in] itemSize The amount of memory per entry in the array.
+ * @param[out] comparator The comparison function to determine if one item is less than another.
+ *
+ * @return The index of the pivot
+ */
+static size_t partition( void * pArray,
+                         size_t low,
+                         size_t high,
+                         size_t itemSize,
+                         ComparisonFunc_t comparator );
+
+/*-----------------------------------------------------------*/
+
+static void swap( void * pFirstItem,
+                  void * pSecondItem,
+                  size_t itemSize )
+{
+    uint8_t * pFirstByte = pFirstItem;
+    uint8_t * pSecondByte = pSecondItem;
+
+    if( ( pFirstItem != NULL ) && ( pSecondItem != NULL ) )
+    {
+        /* Swap one byte at a time. */
+        while( itemSize-- )
+        {
+            uint8_t tmp = *pFirstByte;
+            *pFirstByte++ = *pSecondByte;
+            *pSecondByte++ = tmp;
+        }
+    }
+}
+
+static void quickSortHelper( void * pArray,
+                             size_t low,
+                             size_t high,
+                             size_t itemSize,
+                             ComparisonFunc_t comparator )
+{
+    size_t stack[ high - low + 1 ];
+    /* Low and high are first two items on the stack. */
+    size_t top = 0;
+
+    stack[ top++ ] = low;
+    stack[ top++ ] = high;
+
+    while( top > 0 )
+    {
+        size_t partitionIndex;
+        high = stack[ --top ];
+        low = stack[ --top ];
+
+        partitionIndex = partition( pArray, low, high, itemSize, comparator );
+
+        if( ( partitionIndex != 0U ) && ( partitionIndex - 1U > low ) )
+        {
+            stack[ top++ ] = low;
+            stack[ top++ ] = partitionIndex - 1U;
+        }
+
+        if( partitionIndex + 1U < high )
+        {
+            stack[ top++ ] = partitionIndex + 1U;
+            stack[ top++ ] = high;
+        }
+    }
+}
+
+static size_t partition( void * pArray,
+                         size_t low,
+                         size_t high,
+                         size_t itemSize,
+                         ComparisonFunc_t comparator )
+{
+    void * pivot;
+    size_t i = low - 1U, j = low;
+
+    assert( pArray != NULL );
+
+    pivot = pArray + ( high * itemSize );
```

Comment:
A median of three approach could be used to select the pivot as the median of the first, middle, and last elements

Reply:
That could work but it would also require us to compare strings associated with the field and value for first, middle, and last elements each time the pivot is chosen. That seems slower than just choosing a random pivot.